python/cdo.py

```diff
@@ -754,6 +747,80 @@ def readXDataset(self, ifile=None):
 
     return self.xa_open(ifile)
 
+  def copyNC4Dataset(self, infile, tmpfile):
+    """Creates a temporary file from a netCDF4 Dataset.
+    """
+    ds = self.cdf(tmpfile, mode='w')
+    # copy global attributes
+    for name in infile.ncattrs():
+        ds.setncattr(name, getattr(infile, name))
+    # copy dimensions
+    for name, dimension in infile.dimensions.items():
+        ds.createDimension(
+            name, (len(dimension) if not dimension.isunlimited() else None))
+    # copy data and data attributes
+    for name, variable in infile.variables.items():
+        # _FillValue has to be defined during variable definition
+        if hasattr(variable, '_FillValue'):
+            fill_value = getattr(variable, '_FillValue')
+        else:
+            fill_value = None
+        var = ds.createVariable(name, variable.dtype, variable.dimensions, fill_value=fill_value)
+        for attr in variable.ncattrs():
+            if attr != '_FillValue':
+                ds.variables[name].setncattr(attr, getattr(variable, attr))
+        if variable.shape:
+            ds.variables[name][:] = infile.variables[name][:]
+    ds.close()
+    return tmpfile
+
+  def _input_str(self, input):
+    """Creates an input string from input arguments.
+
+    If input is of type list, each element will be parse by
+    the function _input_arg_to_str. List elements might be of
+    different types.
+    """
+    space = ' '
+    input_str = ''
+    if type(input) == list:
+      for element in input:
+        input_str += space + self._input_arg_to_str(element)
+    else:
+      input_str = space + self._input_arg_to_str(input)
+    return input_str
+
+  def _input_arg_to_str(self, input):
+    """Creates string input element from input argument.
+
+    Input arguments might be any type of filename string,
+    xarray or netcdf4 dataset.
+    """
+    input_str = None
+    if isinstance(input, six.string_types):
+      input_str = input
+    #if type(input) == list:
+    #  input_str = ' '.join(input)
+    # removed elif so we can have netcdf dataset also
+    # when self.hasXarray is true.
+    if self.hasXarray:
```

Comment:
there is now `if` instead of `elif` so that we can also check for netcdf4 dataset even if hasXarray is true.